libmscore/edit.cpp

```diff
@@ -4807,6 +4827,18 @@ void Score::undoAddCR(ChordRest* cr, Measure* measure, const Fraction& tick)
                         toRest(newcr)->setGap(false);
 
                   undo(new AddElement(newcr));
+                  if (newcr->isChord()) {
+                        InstrumentChange* ic = prevInstrumentChange(cr->segment(), staff->idx() * VOICES);
+                        if (ic) {
+                              StaffText* staffText = new StaffText(ic->score());;
+                              if (ic->warning()/* && ic->warning()->parent()*/) {
+                                    undoRemoveElement(ic->warning());
+                              }
+                              ic->setWarning(staffText);
+                              staffText->setPlainText(ic->instrument()->trackName());
+                              newcr->undoAddAnnotation(staffText);
```

Comment:
Would this be better as a function of `InstrumentChange`? Something like `InstrumentChange::setNextEntry(ChordRest* cr)` (or even `undoSetNextEntry`, when you have undo working), which then does all the relevant addition and removal of the actual text element. I just have a feeling it may be needed elsewhere in the future, and either way, since it's dealing with a lot of members of `ic`, it would probably be better off in a function that has direct access to those members.

Reply:
That's a good idea, I think that could remove a lot of code reuse, so I'll refactor that